Docs on `open_virtual_mfdataset` and scaling (#590)
zarr-developers/VirtualiZarr: docs/scaling.md

1. We first must apply the `open_virtual_dataset` function over every file we want to virtualize. This is the map step, and can be parallelized.
2. Then we must take all the resultant virtual datasets (one per file), and combine them together into one final virtual dataset. This is the reduce step.

Finally we write this single virtual dataset to some persistent format.
We have already reduced the data, so this step is a third step, the serialization step.

In our case the amount of data being reduced is fairly small - each virtual dataset is hopefully only a few kBs in memory, small enough to send over the network.
Even a million such virtual datasets together would only require a few GB of RAM in total to hold in memory at once.
This means that as long as we can get all the virtual datasets to be sent back successfully, the reduce step can generally be performed in memory on a single small machine, such as a laptop.
This avoids the need for more complicated parallelization strategies such as a tree-reduce.

## Parallelization Approaches

There are two ways you can implement a map-reduce approach to virtualization in your code.

> The first is to write it yourself, and the second is to use `open_virtual_mfdataset`.

Reviewer: ```suggestion
The first is to write it yourself, and the second is to use [`open_virtual_mfdataset`][virtualizarr.open_virtual_mfdataset].
```

### Manual parallelism

> You are free to call `open_virtual_dataset` on your various files however you like, using any method to apply them, including applying them in parallel.

Reviewer: ```suggestion
You are free to call [`open_virtual_dataset`][virtualizarr.open_virtual_dataset] on your various files however you like, using any method to apply them, including applying them in parallel.
```

For example you may want to parallelize using the [dask library](https://www.dask.org/), which you can do by wrapping each call using `dask.delayed` like this:

```python
import virtualizarr as vz
import dask

tasks = [dask.delayed(vz.open_virtual_dataset)(path) for path in filepaths]
virtual_datasets = dask.compute(tasks)
```

This returns a list of virtual `xr.Dataset` objects, which you can then combine:

```python
import xarray as xr

combined_vds = xr.combine_by_coords(virtual_datasets)
```

### The `parallel` kwarg to `open_virtual_mfdataset`

> Alternatively, you can use `virtualizarr.open_virtual_mfdataset`'s `parallel` keyword argument.

Reviewer: ```suggestion
Alternatively, you can use [virtualizarr.open_virtual_mfdataset][]'s `parallel` keyword argument.
```

This argument allows you to conveniently choose from a range of pre-defined parallel execution frameworks, or even pass your own executor.

The resulting code only takes one function call to generate virtual references in parallel and combine them into one virtual dataset.

```python
combined_vds = vz.open_virtual_mfdataset(filepaths, parallel=<choice_of_executor>)
```

> VirtualiZarr's `open_virtual_mfdataset` is designed to mimic the API of Xarray's `open_mfdataset`, and so accepts all the same keyword argument options for combining.

Reviewer: ```suggestion
VirtualiZarr's [`open_virtual_mfdataset`][virtualizarr.open_virtual_mfdataset] is designed to mimic the API of Xarray's `open_mfdataset`, and so accepts all the same keyword argument options for combining.
```

## Executors

> VirtualiZarr comes with a small selection of executors you can choose from when using `open_virtual_mfdataset`.

Reviewer: ```suggestion
VirtualiZarr comes with a small selection of executors you can choose from when using [`open_virtual_mfdataset`][virtualizarr.open_virtual_mfdataset].
```

!!!note

> If you prefer to do manual parallelism but would like to use one of these executors you can - just import the executor directly from the `virtualizarr.parallel` namespace and use its `.map` method.
> ```

Reviewer: ```suggestion
    If you prefer to do manual parallelism but would like to use one of these executors you can - just import the executor directly from the `virtualizarr.parallel` namespace and use its `.map` method.
```
mkdocs admonitions use intend rather than backticks

### Serial

> The simplest executor is the `SerialExecutor`, which executes all the `open_virtual_dataset` calls in serial, not in parallel.

Reviewer: ```suggestion
The simplest executor is the `SerialExecutor`, which executes all the [`open_virtual_dataset`][virtualizarr.open_virtual_dataset] calls in serial, not in parallel.
```

It is the default executor.

### Threads or Processes

One way to parallelize creating virtual references from a single machine is to across multiple threads or processes.
For this you can use the `ThreadPoolExecutor` or `ProcessPoolExecutor` class from the `concurrent.futures` module in the python standard library.
You simply pass the executor class directly via the `parallel` kwarg.

```python
from concurrent.futures import ThreadPoolExecutor

combined_vds = vz.open_virtual_mfdataset(filepaths, parallel=ThreadPoolExecutor)
```

This can work well when virtualizing files in remote object storage because it parallelizes the issuing of HTTP GET requests for each file.

### Dask Delayed

You can parallelize using `dask.delayed` automatically by passing `parallel='dask'`.

> This will select the `virtualizarr.parallel.DaskDelayedExecutor`.

Reviewer: ```suggestion
This will select the [virtualizarr.parallel.DaskDelayedExecutor][].
```

```python
combined_vds = vz.open_virtual_mfdataset(filepaths, parallel='dask')
```

This uses the same approach that `xarray.open_mfdataset` does when `parallel=True` is passed to it.
Using `dask.delayed` allows for parallelizing with any type of dask cluster, included a managed [Coiled](http://www.coiled.io) cluster.

### Lithops

As the map step is totally embarrassingly parallel, it can be performed entirely using serverless functions.
This approach allows for virtualizing N files in the same time it takes to virtualize 1 file, (assuming you can provision N concurrent serverless functions), avoiding the need to configure, scale, and shutdown a cluster.

You can parallelize VirtualiZarr serverlessly by using the [lithops](http://lithops-cloud.github.io) library.
Lithops can run on all the main cloud provider's serverless FaaS platforms.

> To run on lithops you need to configure lithops for the relevant compute backend (e.g. AWS Lambda), build a runtime using Docker ([example Dockerfile]() with the required dependencies), and ensure the necessary cloud permissions to run are available.

Author: Added in https://github.com/zarr-developers/VirtualiZarr/pull/590/commits/3b4235f4415c065ba6e711c3d769e1aac8be9fb5. 

Will need to be changed to keep up with releases.

> Then you can use the `virtualizarr.parallel.LithopsEagerFunctionExecutor` simply via

Reviewer: ```suggestion
Then you can use the [virtualizarr.parallel.LithopsEagerFunctionExecutor][] simply via:
```

```python
combined_vds = vz.open_virtual_mfdataset(filepaths, parallel='lithops')
```

### Custom Executors

You can also define your own executor to run in some other way, for example on a different serverless platform such as [Modal](https://modal.com).

Your custom executor must inherit from the `concurrent.futures.Executor` ABC, and must implement the `.map` method.

```python
from concurrent.futures import Executor

class CustomExecutor(Executor):
    def map(
        self,
        fn: Callable,
        *iterables: Iterable,
    ) -> Iterator:
        ...

combined_vds = vz.open_virtual_mfdataset(filepaths, parallel=CustomExecutor)
```

## Memory usage

For the virtualization to succeed you need to ensure that your available memory is not exceeded at any point.
There are 3 points at which this might happen:

1. While generating references
2. While combining references
3. While writing references

> While generating references each worker calling `open_virtual_dataset` needs to avoid running out of memory.

Reviewer: ```suggestion
While generating references each worker calling [`open_virtual_dataset`][virtualizarr.open_virtual_dataset] needs to avoid running out of memory.
```

> This primarily depends on how the file is read - see the section on [Caching](scaling.md#Caching) below.

Reviewer: ```suggestion
This primarily depends on how the file is read - see the section on [Caching](#caching-remote-files) below.
```
you can use relative links within a page

> The combine step happens back on the machine on which `vz.open_virtual_mfdataset` was called, so while combining references that machine must have enough memory to hold all the virtual references at once.
> You can find the in-memory size of the references for a single virtual dataset by calling the `.nbytes` accessor method on it (not to be confused with the `.nbytes` xarray method, which returns the total size if all that data were actually loaded into memory).

Reviewer: ```suggestion
The combine step happens back on the machine on which [`open_virtual_mfdataset`][virtualizarr.open_virtual_mfdataset] was called, so while combining references that machine must have enough memory to hold all the virtual references at once.
You can find the in-memory size of the references for a single virtual dataset by calling the [`.nbytes`][virtualizarr.accessor.VirtualiZarrDatasetAccessor.nbytes] accessor method on it (not to be confused with the `.nbytes` xarray method, which returns the total size if all that data were actually loaded into memory).
```

Do this for one file, and multiply by the number of files you have to estimate the total memory required for this step.

Writing the combined virtual references out requires converting them to a different references format, which may have different memory requirements.

## Scalability of references formats

After the map-reduce operation is complete, you will likely still want to persist the virtual references in some format.
Depending on the format, this step may also have scalability concerns.

### Kerchunk

The Kerchunk references specification supports 3 formats - an in-memory (nested) `dict`, JSON, and Parquet.

Both the in-memory Kerchunk `dict` and Kerchunk JSON formats are extremely inefficient ways to represent virtual references.
You may well find that a virtual dataset object that easily fits in memory suddenly uses up many times more memory or space on disk when converted to one of these formats.
Persisting large numbers of references in these formats is therefore not recommended.

> The Kerchunk Parquet format is more scalable, but you may want to experiment with the  `record_size` and `categorical_threshold` arguments to the virtualizarr `.to_kerchunk` accessor method.

Reviewer: ```suggestion
The Kerchunk Parquet format is more scalable, but you may want to experiment with the  `record_size` and `categorical_threshold` arguments to the virtualizarr [`.to_kerchunk`][virtualizarr.accessor.VirtualiZarrDatasetAccessor.to_kerchunk] accessor method.
```

### Icechunk

[Icechunk](https://icechunk.io/) uses it's own [open format](https://icechunk.io/en/latest/spec/) for persisting virtual references.

Icechunk's format stores the virtual references in dedicated binary files, and can use "manifest splitting", together meaning that it should be a scalable way to store large numbers of references.

> TODO: Put numbers on this by testing at large scale once manifest splitting is actually released in Icechunk.

Reviewer: ```suggestion
!!! TODO
    Put numbers on this by testing at large scale once manifest splitting is actually released in Icechunk.
```
Can we put the TODOs in admonitions so they can be more easily skipped over when reading the pages?

## Tips for success

Here are some assorted tips for successfully scaling VirtualiZarr.

### Caching remote files

> When you call `open_virtual_dataset` on a remote file, it  needs to extract the metadata and store it in memory (the returned virtual dataset).

Reviewer: ```suggestion
When you call [`open_virtual_dataset`][virtualizarr.open_virtual_dataset] on a remote file, it  needs to extract the metadata and store it in memory (the returned virtual dataset).
```

One way to do this is to issue HTTP range requests only for each piece of metadata.
This will download the absolute minimum amount of data in total, but issue a lot of HTTP requests, each of which can take a long time to be returned from high-latency object storage.
This approach therefore uses the minimum amount of memory on the worker but takes more time.

> TODO: Describe how this is the default with obstore

Reviewer: ```suggestion
!!! TODO
    Describe how this is the default with obstore
```

## Tips for success

Here are some assorted tips for successfully scaling VirtualiZarr.

### Caching remote files

When you call `open_virtual_dataset` on a remote file, it  needs to extract the metadata and store it in memory (the returned virtual dataset).

One way to do this is to issue HTTP range requests only for each piece of metadata.
This will download the absolute minimum amount of data in total, but issue a lot of HTTP requests, each of which can take a long time to be returned from high-latency object storage.
This approach therefore uses the minimum amount of memory on the worker but takes more time.

TODO: Describe how this is the default with obstore

The other extreme is to download the entire file up front.
This downloads all the metadata by definition, but also all the actual data, which is likely millions of times more than you need for virtualization.
This approach usually takes a lot less time on the worker but requires the maximum amount of memory - using this approach on every file in the dataset entails downloading the entire dataset across all workers!

> TODO: How to enable this by passing `cache=True` to obstore

Author: Requires https://github.com/zarr-developers/VirtualiZarr/issues/625
